docs/decisions/backend/0009-advanced-installer-design.md

```diff
@@ -0,0 +1,246 @@
+
+#### Background Knowledge
+
+Currently Mod Installers work by returning an `IEnumerable<AModFile>`, which is a flat collection of files
+to be moved from one location to another.
+
+To use FOMOD as an example, extracting a file from an individual archive works in the following manner:
+```csharp
+return new FromArchive
+{
+    Id = ModFileId.New(),
+    To = new GamePath(gameTargetPath.Type,
+        gameTargetPath.Path.Join(RelativePath.FromUnsanitizedInput(instruction.destination))),
+    Hash = file.Value.Hash,
+    Size = file.Value.Size
+};
+```
+
+#### Emitting Operations
+
+There may be temptation to dynamically emit the extract (`FromArchive`) instructions as the user selects the install
+target in the UI:
+
+![](./images/0009-advanced-installer-location.png)
+
+This however, is not a good idea; as `undo` operations and unmaps are more difficult to process in a solution like this.
+(Might also be costly in huge mods).
+
+Instead, we should maintain a mapping of `Source` -> `Target`, as a dictionary; and only emit the needed operations 'Just in Time'
+at the deployment stage.
+
+```mermaid
+flowchart TD
+    UD[Update Dictionary] --> EF[Emit FromArchive]
+    EF --> DP[App Deployment]
+```
+
+i.e.
+
+```csharp
+// Note the use of nominal typing.
+public Dictionary<RelativePath, RelativePath> ArchiveToOutputMap;
+```
+
+When the user selects a `target`, we simply add the mapping to the dictionary. Accordingly, we remove the mapping when the
+user undoes the operation.
+
+The implementation should adhere to the [path standards used internally within the Nexus Mods App](./0003-paths.md);
+i.e. All paths should use a separator of `/`, trim whitespace, do not alter `..`, etc.
+
+#### API Design (Proposed)
+
+Note: Initial design, subject to change. Trivial method implementations already included.
+
+```csharp
+/// <summary>
+/// DeploymentData encapsulates all the data needed for a manual or advanced mod installation.
+/// It contains a mapping between the source files in the archive to the target paths in the game directory.
+/// </summary>
+public struct DeploymentData
+{
+    /// <summary>
+    /// Gets or sets the ArchiveToOutputMap property.
+    /// This Dictionary maps relative paths of files in the mod archive to relative paths in the game directory.<br/>
+    /// Key: The relative path of a file within the mod archive.
+    /// Value: The relative path where the file should be placed in the game directory.
+    /// </summary>
+    /// <example>
+    /// If a mod file 'texture.dds' in the archive should go to 'Game/Data/Textures/texture.dds',
+    /// then the KeyValuePair would be like ("texture.dds", "Game/Data/Textures/texture.dds").
+    /// </example>
+    /// <remarks>
+    /// Paths follow internal Nexus Mods App path standards: they use a "/" as a separator, trim whitespace, and do not alter "..".
+    /// </remarks>
+    private Dictionary<RelativePath, RelativePath> _archiveToOutputMap { get; init; } = new();
```

Comment:
Use `Dictionary<RelativePath, GamePath>` instead. We don't want to just target the game directory. 
Some games might need files to be installed in Appdata or Documents/Mygames or somewhere else.
We can map this with GamePaths. 

For now GamePaths can only target 4 locations, defined by the GameFolderType enum, plan is to expand this in the future. For now check which of the GameFolderTypes were defined by the game (or just add them to all games) and show the ones used as suggested folders.

Reply:
Good point, seems I missed this when doublechecking the code.

An earlier version of this actually had the GamePath base path as part of Relative path (in the same way we do for configs). But I decided to scrap that for GamePath, it seems I didn't update this code snippet. Let me do that.